contracts/concentrated/PrivatePool.sol

```diff
@@ -0,0 +1,440 @@
+    event Skim(address indexed provider, uint256[] tokenAmounts);
+
+    constructor(
+        address _owner,
+        address _token0,
+        address _token1
+    ) {
+        factory = msg.sender;
+        owner = _owner;
+        token0 = _token0;
+        token1 = _token1;
+
+        // limit to tokens with decimals <= 18
+        uint256 _token0Decimals = IERC20Metadata(_token0).decimals();
+        require(_token0Decimals <= 18, "token0 decimals > 18");
+        token0Decimals = _token0Decimals;
+
+        uint256 _token1Decimals = IERC20Metadata(_token1).decimals();
+        require(_token1Decimals <= 18, "token1 decimals > 18");
+        token1Decimals = _token1Decimals;
+    }
+
+    /// @notice Updates the quote price LP is willing to offer tokens at
+    /// @param _P The new fixed price LP is willing to buy and sell at
+    function quote(uint256 _P) external onlyOwner {
+        require(_P >= PRICE_MIN && _P <= PRICE_MAX, "price out of range");
+        require(_P != P, "same price");
+
+        // set new P price param
+        P = _P;
+
+        emit Quote(_P);
+    }
+
+    /// @notice Updates the fee applied on swaps
+    /// @dev Effectively acts as bid/ask spread for LP
+    /// @param _fee The new swap fee
+    function setSwapFee(uint256 _fee) external onlyOwner {
+        require(_fee <= FEE_MAX, "fee > max");
+        fee = _fee;
+        emit NewSwapFee(_fee);
+    }
+
+    /// @notice Updates the admin fee applied on private pool swaps
+    /// @dev Admin fees sent to factory owner
+    /// @param _fee The new admin fee
+    function setAdminFee(uint256 _fee) external onlyFactory {
+        require(_fee <= ADMIN_FEE_MAX, "fee > max");
+        adminFee = _fee;
+        emit NewAdminFee(_fee);
+    }
+
+    /// @notice Adds liquidity to pool
+    /// @param amounts The token amounts to add in token decimals
+    /// @param deadline The deadline before which liquidity must be added
+    // TODO: test w protocol fees
+    function addLiquidity(uint256[] memory amounts, uint256 deadline)
+        external
+        onlyOwner
+        deadlineCheck(deadline)
+        hasQuote
+        returns (uint256 minted_)
+    {
+        require(amounts.length == 2, "invalid amounts");
+
+        // get current token balances in pool
+        uint256 xBal = IERC20(token0).balanceOf(address(this));
+        uint256 yBal = IERC20(token1).balanceOf(address(this));
+
+        // convert balances to wad for liquidity calcs adjusted for protocol fees
+        uint256 xWad = _amountWad(xBal - protocolFees0, true);
+        uint256 yWad = _amountWad(yBal - protocolFees1, false);
+
+        // get D balance before add liquidity
+        uint256 _d = _D(xWad, yWad);
+
+        // transfer amounts in decimals in
+        IERC20(token0).safeTransferFrom(msg.sender, address(this), amounts[0]);
+        IERC20(token1).safeTransferFrom(msg.sender, address(this), amounts[1]);
```

Comment:
Thoughts about skipping zero transfers?

Reply:
Think we should probably revert if amount is zero?